Fix qs params without values

miinx opened the pull request:
This adds support to Querystringify for parsing querystring parameters which don't have values, such as `?foo`, & for the reverse scenario: stringifying objects with null value keys.

I've updated the Readme to give examples of both of the above.  

Also included is an update to the `assume` devDependency, as the listed version triggered the npm error described [here](https://github.com/bigpipe/assume/issues/17).

Resolves https://github.com/unshiftio/querystringify/issues/1

lpinca commented on index.js:
```diff
@@ -10,7 +10,7 @@ var has = Object.prototype.hasOwnProperty;
  */
 function querystring(query) {
-  var parser = /([^=?&]+)=([^&]*)/g
+  var parser = /([^=?&]+)=?([^&]*)?/g
```
I think we can remove the final `?`. The second capturing group will be the empty string when there is no value.

miinx replied:
Yep, I love this, wasn't looking at empty string so didn't even try this.  Nice one.  Making the changes now.

lpinca commented on index.js:
```diff
@@ -21,7 +21,7 @@ function querystring(query) {
   for (;
     part = parser.exec(query);
-    result[decodeURIComponent(part[1])] = decodeURIComponent(part[2])
+    result[decodeURIComponent(part[1])] = part[2] ? decodeURIComponent(part[2]) : null
```
I wonder if it's better to keep it as is. When there isn't a value an empty string will be used instead of `null`.
This is consistent with the Node.js querystring module and other popular modules like `qs`:
```js
> querystring.parse('foo&bar=baz')
{ foo: '', bar: 'baz' }
> querystring.parse('foo=&bar=baz')
{ foo: '', bar: 'baz' }
```

miinx replied:
Heh, I looked at a few others (like [query-string](https://www.npmjs.com/package/query-string)) which returned null.  Picked the wrong ones it seems 😄

lpinca commented on index.js:
```diff
@@ -47,7 +48,8 @@ function querystringify(obj, prefix) {
   for (var key in obj) {
     if (has.call(obj, key)) {
-      pairs.push(encodeURIComponent(key) +'='+ encodeURIComponent(obj[key]));
+      value = obj[key] ? '=' + encodeURIComponent(obj[key]) : ''
```
Same thing here. If we use the empty string we can probably leave it as is:
```js
> querystring.stringify({ foo: '', bar: 'baz' })
'foo=&bar=baz'
```

miinx commented:
@lpinca updated the PR.

lpinca commented on README.md:
```diff
@@ -29,14 +29,15 @@ var qs = require('querystringify');
-care if your query string if prefixed with a `?` or not. It just extracts the
-parts between the `=` and `&`:
+The parse method transforms a given query string in to an object. Parameters
+without values are set to `null`. It does not care if your query string is
```
This sentence is no longer true, can you please update it?

lpinca commented on README.md:
```diff
@@ -29,14 +29,15 @@ var qs = require('querystringify');
 qs.parse('?foo=bar');         // { foo: 'bar' }
 qs.parse('foo=bar');          // { foo: 'bar' }
 qs.parse('foo=bar&bar=foo');  // { foo: 'bar', bar: 'foo' }
+qs.parse('foo&bar=foo');      // { foo: null, bar: 'foo' }
```
Please also update these snippets.

lpinca commented on index.js:
```diff
@@ -38,7 +38,8 @@ function querystring(query) {
 
-  var pairs = [];
+  var pairs = []
+    , value;
```
We can remove this variable as it is no longer used.

lpinca commented on index.js:
```diff
@@ -47,7 +48,7 @@ function querystringify(obj, prefix) {
   for (var key in obj) {
     if (has.call(obj, key)) {
-      pairs.push(encodeURIComponent(key) +'='+ encodeURIComponent(obj[key]));
+      pairs.push(encodeURIComponent(key) +'='+ encodeURIComponent(obj[key]))
```
Add back the semicolon.

lpinca commented on test.js:
```diff
@@ -31,6 +31,10 @@ describe('querystringify', function () {
     });
 
+    it('works with object keys with null values', function () {
+      assume(qs.stringify({ foo: '' })).equals('foo=')
```
Missing semicolon.

lpinca commented:
I've added some new comments. Once addressed this pr can be merged.

miinx commented:
@lpinca oops, I shouldn't do things in a rush before work!  Updated as per comments.

One last thing - reverting [this change](https://github.com/unshiftio/querystringify/pull/2/commits/1998c36ca5d5ffa6a0b19ee5d3db2ce5c2f70e21#diff-168726dbe96b3ce427e7fedce31bb0bcL51) earlier meant `stringify` no longer works with null values, only empty string values.  I'm ok with that but wanted to double-check with you?

I've updated the test description [here](https://github.com/unshiftio/querystringify/pull/2/commits/75ab3ee2ccdffa2e8c45d487e235fed814b6bd75#diff-1dd241c4cd3fd1dd89c570cee98b79ddR34) accordingly.

lpinca commented:
> One last thing - reverting this change earlier meant stringify no longer works with null values, only empty string values. I'm ok with that but wanted to double-check with you?

> I've updated the test description here accordingly.

👍

lpinca commented:
Thank you @miinx.
I'm not an owner of the the npm package so I can't publish a new version.
Ping @3rd-Eden.

3rd-Eden commented:
Released as 0.0.4
